### ISSUE TITLE
CLI Feature: State management

### ISSUE THREAD

**Issue by srinjoyc** [onflow/flip-fest#22]
> 👋  &nbsp; If you are interested in working on this issue, please check out the [Getting Started guide on HackerEarth](https://www.hackerearth.com/challenges/hackathon/flip-fest/custom-tab/getting-started/#Getting%20Started)!

## Description (Problem Statement)

The CLI should allow saving the current state and then later restarting from that point similar to how you can use git to move between states. This feature is mostly implemented in the [PR branch](https://github.com/onflow/flow-cli/pull/282) but should be tested, extended with a test suite, and ensure the UX is great.

## Experience Required

- Experience with Go
- Basic understanding of Flow blockchain

## Minimum Feature Set (Acceptance Criteria)

A developer has commands to achieve the following actions:

- saving the state under a name
- restoring to the named state
- deleting a state
- overwriting a state

The state should be persistent between restarts

## Extension (Optional) Feature Set

Interoperability with the Emulator GUI task.

## Milestone Requirements

1. Finish the implementation and making sure it works
2. Write tests around the feature proving the functionality of it
3. Documentation of the feature

## Software Requirements

*Maintainability*

Code should be written with best practices in mind.

*Testing*

Tests should cover all the code implementing this functionality.

## Other Requirements

*Documentation*

- All the commands and the usage should be well documented.

*Code standards or guidelines*

- Follow the general [contribution guidelines](https://github.com/onflow/flow-cli/blob/master/CONTRIBUTING.md).

## Judging Criteria

- For all projects, [technical & high-level judging criteria apply](https://github.com/onflow/flip-fest/blob/main/resources/judging-criteria.md).

## Resources

- [General Resource Guide](https://docs.google.com/document/d/1VXStCyTYRjGNCzTBUW9JAgVenfvLghKq3iMGkMlmroU/edit#heading=h.qpn9u6p0l5y7)
- CLI
    - [Repository](https://github.com/onflow/flow-cli)
    - [Design guidelines](https://github.com/onflow/flow-cli/blob/master/CONTRIBUTING.md#cli-guidelines)
- The [issue](https://github.com/onflow/flow-cli/issues/126)
- [Branch with feature mostly implemented](https://github.com/onflow/flow-cli/pull/282)

**Comment by sideninja** [comment]
Hi 👋
my name is Gregor, and as a software engineer working on developer tools, I will be your point of contact for this task, so good luck, and don't hesitate to reach out to me with any problems or questions. 

I'll keep an eye on your progress and will be reviewing your code.

You can comment here or find me on Discord (I'm sideninja#1970). Join the [Flow Discord server](https://discord.com/invite/J6fFnh2xx6) if you're not there already!

**Comment by bluesign** [comment]
Hey @sideninja , 

I would love to have this feature. I like the snapshot concept, but what would you think to extend this task a bit more to instead emulator to expose subset of git functionality, it would provide a git compatibility layer on storage. 

When new block committed, commit to git state and use Block information, transaction status / events etc as git commit message. 

Git is already providing log, tag, branch, remote origin etc support. Also as cli command same required functionality can be achieved. (just using git as underlying version controlling)

**Comment by sideninja** [comment]
> Hey @sideninja ,
> 
> I would love to have this feature. I like the snapshot concept, but what would you think to extend this task a bit more to instead emulator to expose subset of git functionality, it would provide a git compatibility layer on storage.
> 
> When new block committed, commit to git state and use Block information, transaction status / events etc as git commit message.
> 
> Git is already providing log, tag, branch, remote origin etc support. Also as cli command same required functionality can be achieved. (just using git as underlying version controlling)

That's a very interesting idea. 
Git could be even used as a go module in the emulator to provide this functionality. 
It's a bit funny because the ledger itself has all the history but the storage layer doesn't provide an easy API for this so it probably makes sense to have this kind of "layer 2" solution.

**Comment by bluesign** [comment]
yeah I first tried on ledger side, but branching is a big problem there, moving on a single like is easy, when you try to branch things are getting a bit complicated. 

But with git module it is almost supporting every git operation + cli can translate cli command to git commands too. Also log is pretty nice.

**Comment by bluesign** [comment]
Team: https://www.hackerearth.com/challenges/hackathon/flip-fest/dashboard/3744a70/team/
@bluesign (Tackling solo)
plan time: 2021-10-17

So I will work on this one, I would like to integrate also with https://github.com/onflow/cadence/issues/1160 so we can have a command to use in testing to load, save and switch to some snapshot.

**Comment by bluesign** [comment]
I managed to finish first part: integrating git to emulator, committing to git on transaction executions etc.

But stuck with integrating into cadence ( as discussed in  onflow/cadence#1160 ) 

Actually I managed to integrate by creating a new`ScriptInvocator` and replacing vm's `ScriptProcessors`.  ( Technically switching state in transaction would not be possible anyway )

So far I manage to create a `switchState` cadence built-in method, and by executing script you can change emulator state ( switching to another branch in git ) 

But if someone from FVM side can give some light, maybe there is a better way to do this.

**Comment by kimcodeashian** [comment]
Good day @bluesign!

Thanks so much for all your hardwork & participation. In order to finalize winners & prepare for prize payout, we'll need the following actions from your end.

Please provide the following information by **Nov 17, 2021**,  (in this GH Issue is fine): 

**1. Team Information**
  - Team Members Information - Github Username + Email Contact + Percentage of prize allocation (total should = 100%) 
  - All mentioned members MUST react to the post with a 👍 which will act as confirmation that the information is correct, or a 👎 to indicate that the information is not correct.
  - We will be reaching out via e-mail

**🎖IMPORTANT**: We will only proceed with prize payouts once all members have confirmed with 👍 on the post. 


**2. Video Demo (optional)**
  - Please provide a 5-minute video demo to be featured & showcased in the FLIP Fest Closing Ceremonies
  - Link format & Downloadable (eg. Google Drive, Vimeo)
  - Content Format (Problem Statement, your work / how you solved it, final outcome) 
 

We will be hosting Closing Ceremonies on November 23rd, 8AM PT where we'll having closing remarks from Dete & will be announcing the winners! I'll share the details here before Nov 17.

**Comment by bluesign** [comment]
@bluesign deniz@edincik.com 100%

**Comment by kimcodeashian** [comment]
Hey folks,

We've received and reviewed over **82 submissions**! What an amazing community on Flow! To commemorate all the hard work done, we have finalized winners and will be announcing them during our **[Closing Ceremony](https://forms.gle/j9REJJkVtFFibtfR8) on Nov 23rd, 8AM PT**. Be sure to join us - there may be some attendance prizes & a keynote from our CTO, Dete 😉!  

**[RSVP here](https://forms.gle/j9REJJkVtFFibtfR8)** so you don't miss out! See you then!
